org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/IndexUtils.java

```diff
@@ -0,0 +1,126 @@
+ *******************************************************************************/
+
+package org.eclipse.jdt.ls.core.internal.handlers;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.LinkOption;
+import java.nio.file.StandardCopyOption;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.zip.CRC32;
+
+import org.apache.commons.lang3.StringUtils;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.internal.core.JavaProject;
+import org.eclipse.jdt.internal.core.ClasspathEntry;
+import org.eclipse.jdt.internal.core.index.FileIndexLocation;
+import org.eclipse.jdt.internal.core.index.IndexLocation;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.ProjectUtils;
+
+public class IndexUtils {
+    public static void copyIndexesToSharedLocation() {
+        // common index location for all workspaces
+        final String SHARED_INDEX_LOCATION = System.getProperty("jdt.core.sharedIndexLocation");
+        if (StringUtils.isBlank(SHARED_INDEX_LOCATION)) {
+            return;
+        }
+
+        Set<IndexLocation> reusedIndexLocations = new HashSet<>();
+        IPath javaCoreStateLocation = JavaCore.getPlugin().getStateLocation();
+        for (IJavaProject javaProject : ProjectUtils.getJavaProjects()) {
+            try {
+                IClasspathEntry[] entries = ((JavaProject) javaProject).getResolvedClasspath();
+                for (IClasspathEntry entry : entries) {
+                    if (entry instanceof ClasspathEntry && entry.getEntryKind() == IClasspathEntry.CPE_LIBRARY) {
+                        URL indexURL = ((ClasspathEntry) entry).getLibraryIndexLocation();
+                        if (indexURL == null) {
+                            continue;
+                        }
+
+                        IPath containerPath = entry.getPath();
+                        IndexLocation localIndexLocation = getLocalIndexLocation(javaCoreStateLocation, containerPath);
+                        IndexLocation sharedIndexLocation = IndexLocation.createIndexLocation(indexURL);
+                        if (sharedIndexLocation.equals(localIndexLocation) || reusedIndexLocations.contains(localIndexLocation)) {
+                            continue;
+                        }
+
+                        reusedIndexLocations.add(localIndexLocation);
+                        boolean needDeleteLocalIndex = false;
+                        File localIndexFile = localIndexLocation.getIndexFile();
+                        if (localIndexLocation.exists()) {
+                            File sharedIndexFile = sharedIndexLocation.getIndexFile();
+                            if (sharedIndexFile == null || localIndexFile == null) {
```

Comment:
There is a use case to not skip the file copy.

Each index file contains a header metadata (e.g. `INDEX VERSION 1.131`) at the begining. When the sharedIndexFile is generated by an outdated index scheme (e.g. 1.130), jdt.ls will find it not to match current expected index version, and then regenerate a new version (e.g. 1.131) in local workspace. In such case, we need to overwrite the outdated one in shared location.

I have optimized the code to check if the shared index is being used by current workspace. If so, skip copy and delete local index directly. If not, that means it should be an outdated index and language server will use local index to overwrite it.

Reply:
> Looks like we can check if sharedIndexFile exists on disk. If yes then skip the file copy?

If we support grouping the shared indexes by version, then it's ok to skip the file copy if shared index exists.

I made a PR in jdt.core to support reading shared indexes by version, @rgrunber could you help a review on that? https://github.com/eclipse-jdt/eclipse.jdt.core/pull/564

Reply:
> Looks like we can check if sharedIndexFile exists on disk. If yes then skip the file copy?

@rgrunber This should be the last open question.

Since we cannot tell from file name if the sharedIndexFile contains a correct index version, currently we just always overwrite the sharedIndexFile with local copies. But the disadvantage is that the shared indexes cannot be used by index engine of different versions.

If we skip the file copy when the sharedIndexFile exists, this may result in the shared location containing an outdated version. The indexes are hidden from users, that means if our tools don't keep the index versions update, no one will do that and evetually the shared folder becomes useless.

As a tradeoff, if we save the index versions in different folder, then we can skip file copy if the same shared index version exists.